### ISSUE TITLE
Disable system and user registry for locking

### ISSUE THREAD

**Issue by roberth** [NixOS/nix#12010]
## Is your feature request related to a problem?

- https://github.com/NixOS/nix/issues/12004
- https://github.com/NixOS/nix/issues/7422

## Proposed solution

See https://github.com/NixOS/nix/issues/7422#issuecomment-1341275174
  - Ignore system and user registry in `nix flake lock` / `update`
  - Keep using system and user registry in `nix build <installable>`

## Checklist

- [ ] checked [latest Nix manual] \([source])
- [ ] checked [open feature issues and pull requests] for possible duplicates

[latest Nix manual]: https://nixos.org/manual/nix/unstable/
[source]: https://github.com/NixOS/nix/tree/master/doc/manual/source
[open feature issues and pull requests]: https://github.com/NixOS/nix/labels/feature

---

Add :+1: to [issues you find important](https://github.com/NixOS/nix/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).

**Comment by edolstra** [comment]
Fixed in https://github.com/NixOS/nix/pull/12068.

**Comment by yshui** [comment]
huh this just completely broke my workflow :face_with_diagonal_mouth: with this there is kind of no way to lock a flake with a input refer to the local file system , except by directly writing a full store path in `flake.nix`:

* path to somewhere no in the store: `error: access to absolute path '/xxx' is forbidden in pure evaluation mode`
* `git+file://`: `warning: will not write lock file of flake 'path:/xxx' because it has an unlocked input`
* `path:/nix/store/...`: works

previously i was generating a `registry.json` and referring to it in `flake.nix`, this way updating it can automated. now i have to manually update the store path in `flake.nix`?

**Comment by tanshihaj** [comment]
So now `nix develop <flake-url>#...` and `nix develop` that use flake.nix with `<flake-url>` as an input can generate different environments? Sweet.

Also we are using system registries to allow various nix commands to refer to the exact nixpkgs version this system build from. And now there is no reliable way to do it except patching nix tools to revert changes made in https://github.com/NixOS/nix/pull/12068.

**Comment by roberth** [comment]
@tanshihaj I believe that to be rare, whereas unexpected results from `nix flake lock` were quite common. Hence why we made this change.

> Also we are using system registries to allow various nix commands to refer to the exact nixpkgs version this system build from.

I haven't encountered such a requirement before. What is a use case for it?
Without knowing more details, I'd suggest to either use a single flake for this kind of constraint, or to use the system config flake as an input, and then use `follows` to say something like
```nix
# flake.nix
{
  inputs.config.url = <...>;
  inputs.nixpkgs.follows = "config/nixpkgs";
  outputs = { nixpkgs, ... }: <...>;
}
```

**Comment by tanshihaj** [comment]
> @tanshihaj I believe that to be rare, whereas unexpected results from nix flake lock were quite common. Hence why we made this change.

I do not think that "number of use-cases where it helps more than number of use-cases broken" argument is a good argument since it is still counterintuitive and complicates understanding nix ecosystem. What would you think about python developers if `python -m pytest` would use latest packet from pypi and `import pytest` use one installed you current env or vice versa? 

> I haven't encountered such a requirement before. What is a use case for it?

This becomes really handy when you have a fleet of hosts that are built using different nixpkgs versions and you need to build `mytool` from this nixpkgs of the current system. 

Before the https://github.com/NixOS/nix/pull/12068 you could just add, say, `mynixpkgs` flake in flake registry that points to the nixpkgs version this system is built from. And then both `nix shell mynixpkgs#mytool` and `inputs.nixpkgs.url = "mynixpkgs";` would do the trick. After this commit `nix shell mynixpkgs#mytool` still work fine, **`nix registry` output contains this flake but the flake with  `inputs.nixpkgs.url = "mynixpkgs";` input in this system `cannot find flake 'flake:mynixpkgs' in the flake registries`**. And even worse scenario when you choose `nixpkgs` as this system registry flake - in this case you will get **some** nixpkgs pinned in https://github.com/NixOS/flake-registry instead of one pinned in the system.

I think that even removing both user and system options is better option here. Or create special prefix for the local registries, e.g. `local:nixpkgs` which will mean "I understand what I'm doing, give nixpkgs pinned locally but not in the global registry".

**Comment by yshui** [comment]
> I haven't encountered such a requirement before. What is a use case for it?

For me I have a patched nixpkgs. Instead of maintaining my own fork of nixpkgs and keep everything in sync, I pull upstream nixpkgs and apply a set of patches on them. Then I put the patched nixpkgs in registry (`path: /nix/store/....`), so `<nixpkgs>` can refer to this patched version everywhere, where it's `nix shell`, `nix flake update`, or whatever. This is semi-solvable. Now I override the global flake registry in `nix.conf` (`flake-registry = ...`) and insert my own. (This has the copy from store to store problem, BTW. which hasn't been fully resolved #11229.)

~~Second use case is referring to git repos on local file system, instead of hosting the repo somewhere and go through a `push` `fetch` cycle every time I update it (it is already on my disk, just use it dummy!), I put them in my user registry so `nix` can find it. This one I don't have a workaround, I have to add `--override-flake input url` every time I run `nix flake update`.~~ (ignore this part, this is not accurate.)

@roberth

**Comment by yshui** [comment]
Can't we add a `use-registry = system user` option to `nix.conf` so those of us who want this behavior and don't care about "unexpected results from nix flake lock" can explicitly enable this?

**Comment by roberth** [comment]
> `use-registry = system user`

I'm not a fan but I wouldn't oppose an option like that. It do think it should have something like `lock` in the option name, or make that explicit in some other way.

> What would you think about python developers if python -m pytest would use latest packet from pypi and import pytest use one installed you current env or vice versa?

I'm not a python developer, but IMO the root cause is that all these input declarations are too implicit.
`nix build global:nixpkgs#hello` vs `nix build /etc/nixpkgs#hello` would make these distinctions crystal clear.

> Second use case is referring to git repos on local file system,

This might sound crazy, but I would support a completely general URL rewriting system. The difference is in the framing and usage.
- A registry suggests that this is an application level feature that's meant to be used
- URL rewriting is more easily framed as an optimization, contingency, last resort. Its generality makes it more useful when e.g. certain infrastructure becomes unavailable.

URL rewriting would be a more transient, "configuration and I/O only" affair that leaves fewer traces in the lock. The revision may be incorrect, and therefore still lead to mistakes and errors, but the saved URL remains the same.
So while it has much of the same problems, which are unavoidable given the ask of rewriting, it doesn't seem to unduly legitimize its use.

These issues are of mismanaged expectations, not purely technical.

**Comment by yshui** [comment]
> This might sound crazy, but I would support a completely general URL rewriting system.

Oh, something like Gentoo's [source mirrors](https://www.gentoo.org/downloads/mirrors/)? It's how they fetch source tarballs from alternative URLs. Yeah that sounds like a great idea actually.

To expand on this a bit. I think what you meant is that the lock file only cares about the content hash and shouldn't care where the file is actually fetched from. i.e. it could be fetched from the authoritative source, or the user can specify an alternative mirror URL, as long as the hash is the same it shouldn't matter, and which mirror is used doesn't need to be recorded in the lock file.

**Comment by roberth** [comment]
The way e.g. Nixpkgs sees mirrors is as a list of fallbacks, but for things like substituters (which I think should be included into this rewriting setting), I don't think we want 404s to fall back to the next mirror, because then building something new would take forever.

So while what I had in mind and I believe aligns with the Gentoo approach is "pick one and be happier", Nix could actually support fallback mirrors as well, with the above caveat.
